### DISCUSSION TITLE
org.eclipse.glsp.server.features.sourcemodelwatcher.SourceModelChangedAction.SourceModelChangedAction(String) is not working as expected

### DISCUSSION THREAD

**Discussion by sathya-1994** [eclipse-glsp/glsp#776] [Q&A] [answered]
I am using https://github.com/eclipse-glsp/glsp-vscode-integration project where I have created 3 custom editors for same file extension (3 GLSP diagram editors).

Recently org.eclipse.glsp.server.features.sourcemodelwatcher.SourceModelChangedAction.SourceModelChangedAction(String) was added to close custom editors when file is renamed and it is working fine when we rename from vscode(file rename from explorer view).

But when I trigger SourceModelChangedAction("newFile.wf") from server side then only message "The source model newFile.wf has changed. You might want to consider reloading." will be shown but existing editor will not be closed also new editor with file name newFile.wf is not created.

**When we invoke SourceModelChangedAction from server side then old editors is not closed.**

**Is my understating is correct?**

**My requirement:** I want to close editor and reopen editor with new file name where SourceModelChangedAction is triggered from server side.

**Steps to reproduce the issue:**
1. Inject org.eclipse.glsp.server.actions.ActionDispatcher into org.eclipse.glsp.example.workflow.provider.WorkflowContextMenuItemProvider class

2. Invoke SourceModelChangedAction inside getItems() method(actionDispacther.dispatch(new SourceModelChangedAction("newFile.wf")));

3. Right click on task/automation node.

**Expected:** New editor with file name newFile.wf should we created(old editors should be closed).
**Actual:** Only error message "The source model newFile.wf has changed. You might want to consider reloading."  was shown.

**Comment by planger** [comment, marked as answer]
The [`SourceModelChangedAction`](https://github.com/eclipse-glsp/glsp/blob/master/PROTOCOL.md#244-sourcemodelchangedaction) is an indication, typically sent from the server to the client, that the original source model has changed. From a protocol point of view, clients may react as they see fit though.

The default implementation in the glsp-client is the [`SourceModelChangedActionHandler`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/source-model-watcher/source-model-changed-action-handler.ts). As can be seen in the implementation, the default implementation checks if there is a `ExternalSourceModelChangedHandler` registered. If yes, it'll invoke it, if not, it'll just show a notification.

In Theia, there is a `ExternalSourceModelChangedHandler` registered, which will show a dialog and ask the user whether they want to re-open the editor or not. The implementation is in the [`TheiaSourceModelChangedHandler`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/theia-model-source-changed-handler.ts).

The VS Code integration doesn't register a `ExternalSourceModelChangedHandler` by default. So the editor will just show the notification.

The behavior you are describing in case of a rename is actually coming from plain VS Code:
VS Code reopens the editor with the new file after a rename. This is not something that GLSP does directly whenever a `SourceModelChangedAction` is received.

So if you want a behavior that is similar to the one described above for Theia, you'd need to bind a VS Code specific implementation of the  `ExternalSourceModelChangedHandler`, similar to how it is done for Theia.
Alternatively, you could overwrite the `GlspVscodeConnector` of your example and react to `SourceModelChangedAction`, see also here:
https://github.com/eclipse-glsp/glsp-vscode-integration/blob/d4fccf357cf311fbae0172be1e2767249b99a180/packages/vscode-integration/src/glsp-vscode-connector.ts#L261

As this is pretty generic behavior, please feel free to open PR and add it directly to the default `GlspVscodeConnector` to. We'd welcome your contribution!
